nats/src/main.rs

```diff
@@ -168,7 +232,10 @@ impl NatsMessagingProvider {
             // up with 20 subscriptions all getting slammed with up to 75 tasks, but we should wait
             // to do anything until we see what happens with real world usage and benchmarking
             let semaphore = Arc::new(Semaphore::new(75));
-            while let Some(msg) = subscription.next().await {
+
+            let mut shared_subscriber = shared_subscriber.write().await;
+
```

Comment:
I think we have to be careful here, the loop below should run for the full length of the provider on behalf of a single subscription, so if we need to get a lock on this subscriber elsewhere.

E.g. I'm concerned this write lock will be held even when we try and clean up the subscription. Just want to verify with testing that's not the case

Reply:
Yeah thanks for noting this -- considering that `delete_link` takes the lock to `actors` (and I didn't introduce any new locks there I assumed that this would be fine... 

What would be the best way to go about testing this? It seems like the tests in `nats_test.rs` would cover this but I can write an explicit test case.

Reply:
I think Brooks is right, and as written this code won't allow the provider to close connections when a link is deleted:

- `put_link` calls `connect`, which calls this `subscribe` function
- we spawn a new tokio task which acquires the write lock on `shared_subscriber`. This function runs "forever" (unless the NATS server closes the subscription, or there's an error in the client)
- `delete_link` tries to acquire a write lock on the same subscriber. Since this lock is already being held, we'll never be able to call `unsubscribe`

Reply:
looking at it, I'm not sure there's a way the listener stops other than by being overwhelmed with messages -- is that right?

I *think* I might have found a way to proivde a spot for the lock to be released:

```rust
            while let Some(msg) = shared_subscriber.write().await.next().await {
                ...
            }
```

Looks like also this is a [[well known dark corner](https://github.com/rust-lang/book/issues/1871) -- I'll switch to `loop`here and explicitly drop, but not quite sure how this will affect perf.

Reply:
Ah sorry @connorsmith256 didn't see your message there -- yeah looking at it, I  wasn't aware that the loop ran for the entire time/didn't stop -- I've replaced the outer shared sub lock acquisition with a `loop` and explicit `drop` to give other writers a chance

Reply:
This has been refactored to directly use JoinHandles